modules/platform/nuxeo-platform-document-routing/nuxeo-routing-core/src/main/java/org/nuxeo/ecm/platform/routing/core/impl/DocumentRoutingServiceImpl.java

```diff
@@ -143,51 +139,58 @@
      */
     public static final String ACTOR_ACE_CREATOR = "Workflow";
 
-    // FIXME: use ContributionFragmentRegistry instances instead to handle hot
-    // reload
-
     public static final String ROUTE_MODELS_IMPORTER_XP = "routeModelImporter";
 
-    protected Map<String, String> typeToChain = new HashMap<>();
-
-    protected Map<String, String> undoChainIdFromRunning = new HashMap<>();
+    protected RepositoryInitializationHandler repositoryInitializationHandler = new RouteModelsInitializator();
 
-    protected Map<String, String> undoChainIdFromDone = new HashMap<>();
+    private Cache<String, String> modelsCache;
 
     protected DocumentRoutingPersister persister;
 
-    protected RouteTemplateResourceRegistry routeResourcesRegistry = new RouteTemplateResourceRegistry();
+    protected List<URL> routeTemplates;
 
-    protected RepositoryInitializationHandler repositoryInitializationHandler;
+    protected Map<String, String> typeToChain;
 
-    private Cache<String, String> modelsChache;
+    protected Map<String, String> undoChainIdFromRunning;
+
+    protected Map<String, String> undoChainIdFromDone;
 
     @Override
-    public void registerContribution(Object contribution, String extensionPoint, ComponentInstance contributor) {
-        if (CHAINS_TO_TYPE_XP.equals(extensionPoint)) {
-            ChainToTypeMappingDescriptor desc = (ChainToTypeMappingDescriptor) contribution;
-            typeToChain.put(desc.getDocumentType(), desc.getChainId());
-            undoChainIdFromRunning.put(desc.getDocumentType(), desc.getUndoChainIdFromRunning());
-            undoChainIdFromDone.put(desc.getDocumentType(), desc.getUndoChainIdFromDone());
-        } else if (PERSISTER_XP.equals(extensionPoint)) {
-            PersisterDescriptor des = (PersisterDescriptor) contribution;
+    public void start(ComponentContext context) {
+        modelsCache = CacheBuilder.newBuilder().maximumSize(100).expireAfterWrite(10, TimeUnit.MINUTES).build();
+        this.<PersisterDescriptor> getRegistryContribution(PERSISTER_XP).ifPresent(desc -> {
             try {
-                persister = des.getKlass().getDeclaredConstructor().newInstance();
+                persister = desc.getKlass().getDeclaredConstructor().newInstance();
             } catch (ReflectiveOperationException e) {
                 throw new NuxeoException(e);
             }
-        } else if (ROUTE_MODELS_IMPORTER_XP.equals(extensionPoint)) {
-            RouteModelResourceType res = (RouteModelResourceType) contribution;
-            registerRouteResource(res, contributor.getRuntimeContext());
-        }
+        });
+        routeTemplates = getRegistryContributions(ROUTE_MODELS_IMPORTER_XP).stream()
+                                                                           .map(RouteModelResourceType.class::cast)
```

Comment:
Could avoid cast by doing `this.<RouteModelResourceType> getRegistryContributions` ?